procrastinate/testing.py

```diff
@@ -94,55 +103,59 @@ async def close_async(self) -> None:
 
     # End of BaseConnector methods
 
-    async def defer_job_one(
-        self,
-        task_name: str,
-        priority: int,
-        lock: str | None,
-        queueing_lock: str | None,
-        args: types.JSONDict,
-        scheduled_at: datetime.datetime | None,
-        queue: str,
-    ) -> JobRow:
-        if queueing_lock is not None and any(
-            job["queueing_lock"] == queueing_lock and job["status"] == "todo"
-            for job in self.jobs.values()
-        ):
-            from . import manager
+    async def defer_jobs_all(self, jobs: list[types.JobToDefer]) -> list[JobRow]:
+        # We check the queueing locks upfront so that no job is inserted into
+        # the queue if the constraint is violated (simulating a database
+        # rollback).
+        seen_queueing_locks = set()
+        for job in jobs:
+            queueing_lock = job.queueing_lock
+            if queueing_lock is not None and (
+                any(
+                    job["queueing_lock"] == queueing_lock and job["status"] == "todo"
+                    for job in self.jobs.values()
+                )
+                or queueing_lock in seen_queueing_locks
+            ):
+                raise exceptions.UniqueViolation(
+                    constraint_name=manager.QUEUEING_LOCK_CONSTRAINT
+                )
+
+            seen_queueing_locks.add(queueing_lock)
```

Comment:
```suggestion
        current_queuing_locks = {
            job["queueing_lock"]
            for job in self.jobs.values()
            if job["status"] == "todo"
        } - {"None}
        
        new_queueing_locks = [job.queueing_lock for job in jobs]
        new_queueing_locks_set = set(new_queueing_locks)
        if (
            len(new_queueing_locks) != len(new_queueing_locks_set)
            or current_queuing_locks & current_queuing_locks
        ):
                raise exceptions.UniqueViolation(
                    constraint_name=manager.QUEUEING_LOCK_CONSTRAINT
                )
```

Not sure it's more readable, what do you think ?